### PR TITLE
more informative GCPNodeProvider create_node return

### PR DESCRIPTION
**Pull request by hanming-lu**
## Why are these changes needed?
- More informative return value for GCPNodeProvider create_node

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/gcp/node_provider.py [hunk 1]
```diff
@@ -164,14 +164,17 @@ def internal_ip(self, node_id: str):
             return ip
 
     @_retry
-    def create_node(self, base_config: dict, tags: dict, count: int) -> None:
+    def create_node(
+        self, base_config: dict, tags: dict, count: int
+    ) -> List[Tuple[dict, str]]:
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/gcp/node_provider.py:169]: Thanks for the PR!

Could you add a doc string explaining the meaning of the returned data?
Also, could you modify the return type on the create node method in the NodeProvider interface to `Any` and change the docstring there to reflect the more general return type?
https://github.com/ray-project/ray/blob/905258dbc19753c81039f993477e7ab027960729/python/ray/autoscaler/node_provider.py#L121-L124

Comment (hanming-lu) [reply python/ray/autoscaler/_private/gcp/node_provider.py]: Sounds good! I made some changes to the return value to make it consistent with NodeProvider and AWSNodeProvider, as well as added a doc string. Let me know if it looks better for you.

File: python/ray/autoscaler/_private/gcp/node_provider.py [hunk 2]
```diff
@@ -164,14 +164,23 @@ def internal_ip(self, node_id: str):
...
             labels = tags  # gcp uses "labels" instead of aws "tags"
 
             node_type = get_node_type(base_config)
             resource = self.resources[node_type]
 
-            resource.create_instances(base_config, labels, count)
+            results = resource.create_instances(base_config, labels, count)
+            instance_dict: Dict[str, None] = {}
+            for _, instance_id in results:
+                instance_dict[instance_id] = None
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/gcp/node_provider.py:182]: eh...None feels a little weird.
Can we actually use the response dict (_) as the value?

Comment (hanming-lu) [reply python/ray/autoscaler/_private/gcp/node_provider.py]: Good point. Have just updated that :)

**Review by DmitriGekhtman — APPROVED** [review]
Looks good, thanks!

**Comment by DmitriGekhtman** [comment]
Out of abundance of caution (the code involved is not tested), I launched a head node using this PR branch.

File: python/ray/autoscaler/_private/gcp/node_provider.py [hunk 3]
```diff
@@ -164,14 +164,23 @@ def internal_ip(self, node_id: str):
             return ip
 
     @_retry
-    def create_node(self, base_config: dict, tags: dict, count: int) -> None:
+    def create_node(self, base_config: dict, tags: dict, count: int) -> Dict[str, dict]:
+        """Creates instances.
+
+        Returns dict mapping instance id to each create operation result for the created
+        instances.
+        """
```
Comment (cadedaniel) [inline python/ray/autoscaler/_private/gcp/node_provider.py:172]: Thanks for improving the documentation!

**Review by cadedaniel — APPROVED** [review]

File: python/ray/autoscaler/_private/gcp/node_provider.py [hunk 4]
```diff
@@ -164,14 +164,23 @@ def internal_ip(self, node_id: str):
...
 
             node_type = get_node_type(base_config)
             resource = self.resources[node_type]
 
-            resource.create_instances(base_config, labels, count)
+            results = resource.create_instances(base_config, labels, count)
+            instance_dict: Dict[str, dict] = {}
+            for result, instance_id in results:
+                instance_dict[instance_id] = result
+            return instance_dict
```
[the remark this answers is not in the document]
Comment (cadedaniel) [reply python/ray/autoscaler/_private/gcp/node_provider.py:183]: Also I'm concerned that `for result, instance_id in results:` will throw `ValueError`. How does this work? See the following REPL code. I think you'll need to add `results.items()`.
```python3
results = {'key1': 'value1', 'key2': 'value2'}
>>> for k, v in results:
...     print(k, v)
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: too many values to unpack (expected 2)
```

**Review by cadedaniel — DISMISSED** [review]
~I think the for loop will throw `ValueError`~ requesting that we add a unit test.

File: python/ray/autoscaler/_private/gcp/node_provider.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cadedaniel) [reply python/ray/autoscaler/_private/gcp/node_provider.py:183]: @DmitriGekhtman do we have no way to write a unit test for this?

[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/gcp/node_provider.py:183]: @cadedaniel `results` is a list of tuples

[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/gcp/node_provider.py:183]: > @DmitriGekhtman do we have no way to write a unit test for this?

We're hoping you can figure that out for us :)

Here are the existing unit tests:
https://github.com/ray-project/ray/blob/master/python/ray/tests/gcp/test_gcp_node_provider.py
https://github.com/ray-project/ray/blob/master/python/ray/tests/test_autoscaler_gcp.py

[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/gcp/node_provider.py:183]: @hanming-lu I think for extra clarity and over-documentation (never a bad thing) I think we can add a type annotation to "results" in the line
`results = resource.create_instances(base_config, labels, count)`

[the remark this answers is not in the document]
Comment (cadedaniel) [reply python/ray/autoscaler/_private/gcp/node_provider.py:183]: > @cadedaniel results is a list of tuples, so we're good on this

Oh, sweet! Got it.


I think we should have unit tests for this change. It isn't that difficult to mock, here's some starter code:
```python3
def test_create_node_returns_dict():
  mock_results = [('a', 'b'), ('c', 'd')]
  mock_resource = MagicMock()
  mock_resource.create_instances.return_value = mock_results
  def __init__(self, provider_config: dict, cluster_name: str):
    self.lock = RLock()
    self.cached_nodes: Dict[str, GCPNode] = {}
    self.resources[<mock node type here>] = mock_resource
  with patch.object(GCPNodeProvider, '__init__', __init__):
    node_provider = GCPNodeProvider(...)
    create_node_return_value = node_provider.create_instances(...)
  assert create_node_return_value == {'b': 'a', 'd': 'c'}
```

**Review by DmitriGekhtman — CHANGES_REQUESTED** [review]
A unit test, as suggested by Cade, would be great.

File: python/ray/autoscaler/_private/gcp/node_provider.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanming-lu) [reply python/ray/autoscaler/_private/gcp/node_provider.py]: Sounds good, I have added both inline type annotation and a unit test :)

File: python/ray/autoscaler/_private/gcp/node_provider.py [hunk 5]
```diff
@@ -164,14 +164,23 @@ def internal_ip(self, node_id: str):
...
+        instances.
+        """
         with self.lock:
             labels = tags  # gcp uses "labels" instead of aws "tags"
 
             node_type = get_node_type(base_config)
             resource = self.resources[node_type]
 
-            resource.create_instances(base_config, labels, count)
+            results = resource.create_instances(base_config, labels, count)
```
[the remark this answers is not in the document]
Comment (hanming-lu) [reply python/ray/autoscaler/_private/gcp/node_provider.py]: Sounds good, I have added both inline type annotation and a unit test :)

**Review by DmitriGekhtman — APPROVED** [review]

**Review by cadedaniel — APPROVED** [review]

File: python/ray/autoscaler/_private/gcp/node_provider.py [hunk 6]
```diff
@@ -164,14 +164,23 @@ def internal_ip(self, node_id: str):
...
 
             node_type = get_node_type(base_config)
             resource = self.resources[node_type]
 
-            resource.create_instances(base_config, labels, count)
+            results = resource.create_instances(base_config, labels, count)  # type: List[Tuple[dict, str]] 
+            instance_dict: Dict[str, dict] = {}
+            for result, instance_id in results:
+                instance_dict[instance_id] = result
+            return instance_dict
```
Comment (cadedaniel) [inline python/ray/autoscaler/_private/gcp/node_provider.py:183]: Hey @hanming-lu, did you see my suggestion? We're trying to improve craftsmanship of our code, and while what you have is good, further simplification is even better. What do you think of removing this for loop and variable creation to this single dict comprehension?
```python
return {instance_id: result for result, instance_id in results}
```

Comment (hanming-lu) [reply python/ray/autoscaler/_private/gcp/node_provider.py]: My apologies, I missed the comment. I have just updated it :)

**Comment by DmitriGekhtman** [comment]
There was a lint failure (a very [weird one](https://buildkite.com/ray-project/ray-builders-pr/builds/34527#01813b44-be5d-4a22-9431-26d8e738de14) at that).
You might be able to fix it by running the lint hooks locally. There's some info on the hooks here:
https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#submitting-and-merging-a-contribution

**Comment by DmitriGekhtman** [comment]
Test failures unrelated.

**Pull request merged**